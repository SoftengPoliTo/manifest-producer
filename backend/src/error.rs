use std::error::Error as StdError;
use std::fmt;

#[derive(Debug)]
pub enum Error {
    Io(std::io::Error),
    InvalidFormat(goblin::error::Error),
    InvalidFileName,
    Capstone(capstone::Error),
    DebugInfo,
    LangNotFound,
    TextSectionNotFound,
    DemanglingError(cpp_demangle::error::Error),
    FormattingError(std::fmt::Error),
    Json(serde_json::Error),
    GimliError(gimli::Error),
    ObjectError(object::Error),
    FunctionNotFound(String),
    ProgressStyleError(indicatif::style::TemplateError),
}

impl Error {
    pub fn description(&self) -> &str {
        match self {
            Error::Io(_) => "I/O error",
            Error::InvalidFormat(_) => "Invalid format file",
            Error::InvalidFileName => "No valid file name",
            Error::Capstone(_) => "Capstone error",
            Error::DebugInfo => "No Debug-info",
            Error::LangNotFound => "Language not found",
            Error::TextSectionNotFound => ".text section not found",
            Error::DemanglingError(_) => "Demangling error",
            Error::FormattingError(_) => "Formatting error",
            Error::Json(_) => "JSON serialization error",
            Error::GimliError(_) => "Gimli error",
            Error::ObjectError(_) => "Object error",
            Error::FunctionNotFound(_) => "Function not found",
            Error::ProgressStyleError(_) => "Progress style error",
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::Io(e) => write!(f, "I/O error -> {}", e),
            Error::InvalidFormat(msg) => write!(f, "Invalid format file -> {}", msg),
            Error::InvalidFileName => write!(f, "No valid file name"),
            Error::Capstone(e) => write!(f, "Capstone error -> {}", e),
            Error::DebugInfo => write!(f, "No debug info available"),
            Error::LangNotFound => write!(f, "Programming language not found"),
            Error::TextSectionNotFound => write!(f, "'.text' section not found"),
            Error::DemanglingError(e) => write!(f, "Demangling error -> {}", e),
            Error::FormattingError(e) => write!(f, "Formatting error -> {}", e),
            Error::Json(e) => write!(f, "JSON serialization error -> {}", e),
            Error::GimliError(e) => write!(f, "Gimli error -> {}", e),
            Error::ObjectError(e) => write!(f, "Object error -> {}", e),
            Error::FunctionNotFound(func) => write!(f, "Function '{}' not found", func),
            Error::ProgressStyleError(e) => write!(f, "Progress style error -> {}", e),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        match self {
            Error::Io(e) => Some(e),
            Error::InvalidFormat(e) => Some(e),
            Error::Capstone(e) => Some(e),
            Error::DemanglingError(e) => Some(e),
            Error::FormattingError(e) => Some(e),
            Error::Json(e) => Some(e),
            Error::GimliError(e) => Some(e),
            Error::ObjectError(e) => Some(e),
            Error::ProgressStyleError(e) => Some(e),
            _ => None,
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(err: std::io::Error) -> Self {
        Error::Io(err)
    }
}

impl From<goblin::error::Error> for Error {
    fn from(err: goblin::error::Error) -> Self {
        Error::InvalidFormat(err)
    }
}

impl From<capstone::Error> for Error {
    fn from(err: capstone::Error) -> Self {
        Error::Capstone(err)
    }
}

impl From<cpp_demangle::error::Error> for Error {
    fn from(err: cpp_demangle::error::Error) -> Self {
        Error::DemanglingError(err)
    }
}

impl From<std::fmt::Error> for Error {
    fn from(err: std::fmt::Error) -> Self {
        Error::FormattingError(err)
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Error::Json(err)
    }
}

impl From<gimli::Error> for Error {
    fn from(err: gimli::Error) -> Self {
        Error::GimliError(err)
    }
}

impl From<object::Error> for Error {
    fn from(err: object::Error) -> Self {
        Error::ObjectError(err)
    }
}

impl From<indicatif::style::TemplateError> for Error {
    fn from(err: indicatif::style::TemplateError) -> Self {
        Error::ProgressStyleError(err)
    }
}

pub type Result<T> = std::result::Result<T, Error>;
