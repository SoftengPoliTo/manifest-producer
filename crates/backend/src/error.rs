use std::error::Error as StdError;
use std::fmt;

#[derive(Debug)]
pub enum Error {
    Io(std::io::Error),
    InvalidFormat(goblin::error::Error),
    InvalidFileName,
    Capstone(capstone::Error),
    DebugInfo,
    LangNotFound,
    TextSectionNotFound,
    DemanglingError(cpp_demangle::error::Error),
    FormattingError(std::fmt::Error),
    Json(serde_json::Error),
    GimliError(gimli::Error),
    ObjectError(object::Error),
    FunctionNotFound(String),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::Io(e) => write!(f, "I/O error -> {e}"),
            Error::InvalidFormat(msg) => write!(f, "Invalid format file -> {msg}"),
            Error::InvalidFileName => write!(f, "No valid file name"),
            Error::Capstone(e) => write!(f, "Capstone error -> {e}"),
            Error::DebugInfo => write!(f, "No debug info available"),
            Error::LangNotFound => write!(f, "Programming language not found"),
            Error::TextSectionNotFound => write!(f, "'.text' section not found"),
            Error::DemanglingError(e) => write!(f, "Demangling error -> {e}"),
            Error::FormattingError(e) => write!(f, "Formatting error -> {e}"),
            Error::Json(e) => write!(f, "JSON serialization error -> {e}"),
            Error::GimliError(e) => write!(f, "Gimli error -> {e}"),
            Error::ObjectError(e) => write!(f, "Object error -> {e}"),
            Error::FunctionNotFound(func) => write!(f, "Function '{func}' not found"),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        match self {
            Error::Io(e) => Some(e),
            Error::InvalidFormat(e) => Some(e),
            Error::Capstone(e) => Some(e),
            Error::DemanglingError(e) => Some(e),
            Error::FormattingError(e) => Some(e),
            Error::Json(e) => Some(e),
            Error::GimliError(e) => Some(e),
            Error::ObjectError(e) => Some(e),
            _ => None,
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(err: std::io::Error) -> Self {
        Error::Io(err)
    }
}

impl From<goblin::error::Error> for Error {
    fn from(err: goblin::error::Error) -> Self {
        Error::InvalidFormat(err)
    }
}

impl From<capstone::Error> for Error {
    fn from(err: capstone::Error) -> Self {
        Error::Capstone(err)
    }
}

impl From<cpp_demangle::error::Error> for Error {
    fn from(err: cpp_demangle::error::Error) -> Self {
        Error::DemanglingError(err)
    }
}

impl From<std::fmt::Error> for Error {
    fn from(err: std::fmt::Error) -> Self {
        Error::FormattingError(err)
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Error::Json(err)
    }
}

impl From<gimli::Error> for Error {
    fn from(err: gimli::Error) -> Self {
        Error::GimliError(err)
    }
}

impl From<object::Error> for Error {
    fn from(err: object::Error) -> Self {
        Error::ObjectError(err)
    }
}

pub type Result<T> = std::result::Result<T, Error>;
