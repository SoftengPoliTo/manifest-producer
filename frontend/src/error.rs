use std::error::Error as StdError;
use std::fmt;

/// Custom error type for the `frontend` crate.
///
/// This enum encapsulates various error types that may occur in the frontend module:
/// - `Io`: Errors related to file input/output operations.
/// - `Json`: Errors encountered during JSON serialization or deserialization.
/// - `ProgressStyleError`: Errors in styling or formatting progress indicators using the `indicatif` crate.
/// - `MiniJinjaError`: Errors raised by the `minijinja` templating engine.
#[derive(Debug)]
pub enum Error {
    Io(std::io::Error),
    Json(serde_json::Error),
    ProgressStyleError(indicatif::style::TemplateError),
    MiniJinjaError(minijinja::Error),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::Io(e) => write!(f, "I/O error -> {}", e),
            Error::Json(e) => write!(f, "JSON serialization error -> {}", e),
            Error::ProgressStyleError(e) => write!(f, "Progress style error -> {}", e),
            Error::MiniJinjaError(e) => write!(f, "MiniJinja error -> {}", e),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        match self {
            Error::Io(e) => Some(e),
            Error::Json(e) => Some(e),
            Error::ProgressStyleError(e) => Some(e),
            Error::MiniJinjaError(e) => Some(e),
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(err: std::io::Error) -> Self {
        Error::Io(err)
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Error::Json(err)
    }
}

impl From<indicatif::style::TemplateError> for Error {
    fn from(err: indicatif::style::TemplateError) -> Self {
        Error::ProgressStyleError(err)
    }
}

impl From<minijinja::Error> for Error {
    fn from(err: minijinja::Error) -> Self {
        Error::MiniJinjaError(err)
    }
}

pub type Result<T> = std::result::Result<T, Error>;
