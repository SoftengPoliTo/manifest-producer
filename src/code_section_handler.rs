use std::collections::HashMap;

use goblin::elf::Elf;

use crate::{
    elf_utils::{cs_init, find_text_section, get_name_addr, API},
    error,
    plt_mapping::{find_plt_section, load_rela_plt_relocations},
};
use error::{Error, Result};

/// Extracts and disassembles code sections of APIs, handling static or dynamic linking.
///
/// This function extracts and disassembles the code of the specified API section, managing static or dynamic linking.
/// It returns the disassembled code and any system calls made by the API.
///
/// # Arguments
///
/// * `elf` - The ELF object representing the binary.
/// * `api` - The API structure containing information about the API section.
/// * `buffer` - The buffer containing the binary data of the ELF file.
/// * `link` - A boolean indicating whether static linking is used (`true`) or dynamic linking (`false`).
/// * `rust` - A boolean indicating whether the API section is written in Rust (`true`) or not (`false`).
///
/// # Returns
///
/// Returns a `Result` containing a vector of disassembled code strings and any system calls made by the API.
pub fn code_section(
    elf: &Elf,
    api: &API,
    buffer: &[u8],
    link: bool,
    rust: bool,
) -> Result<Vec<String>> {
    let text_section = find_text_section(elf).ok_or(Error::TextSectionNotFound)?;
    let code_slice: &[u8];
    let sys_call;

    if link {
        // Static linking
        let text_start_index = text_section.sh_offset as usize;
        let func_start_offset = (api.start_addr - text_section.sh_addr) as usize;
        let func_end_offset = (api.end_addr - text_section.sh_addr) as usize;
        code_slice =
            &buffer[text_start_index + func_start_offset..text_start_index + func_end_offset];

        println!("\n{:#x}\t<{}>", &api.start_addr, &api.name);
        sys_call = disassemble(elf, code_slice, api.start_addr, link, None, rust)?;
    } else {
        // Dynamic linking
        code_slice = &buffer[(api.start_addr) as usize..(api.end_addr) as usize];

        let mut found_plt_sec = false;
        let plt_section =
            find_plt_section(elf, &mut found_plt_sec).ok_or(Error::PLTSectionNotFound)?;
        let plt_entry_size = plt_section.sh_entsize as usize;
        let tbl = load_rela_plt_relocations(elf, plt_section, plt_entry_size, found_plt_sec);

        println!("\n{:#x}\t<{}>", &api.start_addr, &api.name);
        sys_call = disassemble(elf, code_slice, api.start_addr, link, tbl, rust)?;
    }

    Ok(sys_call)
}

// Disassembles the code in the specified section, handling static or dynamic function calls.
//
// This function disassembles the code in the specified section, handling static or dynamic function calls based on the given parameters.
// It returns the disassembled code and any system calls made by the API.
fn disassemble(
    elf: &Elf,
    code_slice: &[u8],
    addr: u64,
    link: bool,
    plt_map: Option<HashMap<u64, &str>>,
    rust: bool,
) -> Result<Vec<String>> {
    let cs = cs_init()?;
    let mut sys_call: Vec<String> = vec![];

    let instructions = cs.disasm_all(code_slice, addr).unwrap();
    for insn in instructions.iter() {
        let insn_addr = insn.address();
        let insn_name = cs.insn_name(insn.id()).unwrap();
        let op_str = insn.op_str().unwrap();

        if rust && insn_name == "lea" {
            if let Some(name) = lea_instruction(elf, op_str, insn_addr, insn_name.clone()) {
                sys_call.push(name);
            }
        } else if insn_name == "call" && !rust {
            if let Some(name) = call_instruction(
                elf,
                op_str,
                insn_addr,
                insn_name.clone(),
                link,
                plt_map.clone(),
            ) {
                sys_call.push(name);
            }
        } else {
            println!("0x{:x}:\t{}\t{}", insn_addr, insn_name, op_str);
        }
    }
    Ok(sys_call)
}

// Handles the instruction 'lea', identifies the function name, and adds any interface called by API.
fn lea_instruction<'a>(
    elf: &'a Elf<'a>,
    op_str: &'a str,
    insn_addr: u64,
    insn_name: String,
) -> Option<String> {
    if let Some(offset_str) = op_str.strip_suffix("(%rip), %rax") {
        if offset_str.starts_with('-') {
            if let Some(addr_str) = offset_str.strip_prefix("-0x") {
                if let Ok(addr) = u64::from_str_radix(addr_str, 16) {
                    let target_addr = insn_addr.wrapping_sub(addr);
                    let target_addr_aligned = target_addr + 7;
                    if let Some(name) = get_name_addr(elf, target_addr_aligned) {
                        println!("0x{:x}:\t{}\t<{}>", insn_addr, insn_name, name);
                        return Some(name.to_string());
                    }
                }
            }
        } else if let Some(addr_str) = offset_str.strip_prefix("0x") {
            if let Ok(addr) = u64::from_str_radix(addr_str, 16) {
                let target_addr = insn_addr.wrapping_add(addr);
                let target_addr_aligned = target_addr + 7;
                if let Some(name) = get_name_addr(elf, target_addr_aligned) {
                    println!("0x{:x}:\t{}\t<{}>", insn_addr, insn_name, name);
                    return Some(name.to_string());
                }
            }
        }
    }
    None
}

// Handles the instruction 'call', identifies the function name, and adds any interface called by API.
fn call_instruction<'a>(
    elf: &'a Elf<'a>,
    op_str: &'a str,
    address: u64,
    name_func: String,
    link: bool,
    plt_map: Option<HashMap<u64, &str>>,
) -> Option<String> {
    if let Some(addr_str) = op_str.strip_prefix("0x") {
        if let Ok(addr) = u64::from_str_radix(addr_str, 16) {
            if link {
                // Statically linked
                if let Some(name) = get_name_addr(elf, addr) {
                    println!("0x{:x}:\t{}\t<{}>", address, name_func, name);
                    return Some(name.to_string());
                }
                let name = format!("CALL_to_<{}>", op_str);
                println!("0x{:x}:\t{}\t<{}>", address, name_func, name);
                return Some(name.to_string());
            }
            // Dynamically linked
            match plt_map {
                Some(map) => {
                    if let Some(plt_value) = map.get(&addr) {
                        println!("0x{:x}:\t{}\t<{}>", address, name_func, plt_value);
                        return Some(plt_value.to_string());
                    }
                    if let Some(name) = get_name_addr(elf, addr) {
                        println!("0x{:x}:\t{}\t<{}>", address, name_func, name);
                        return Some(name.to_string());
                    }
                }
                None => {
                    println!("PLT map is not available");
                }
            }
        } else {
            println!("Invalid address format: {}", op_str);
        }
    }
    None
}
