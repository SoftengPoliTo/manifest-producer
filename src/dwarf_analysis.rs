use crate::error;
use error::Result;
use std::{fs, borrow};
use object::{Object, ObjectSection};

/// Parse an ELF file to determine the programming language used.
///
/// This function analyzes the Dwarf information in the ELF file to determine the programming language used.
///
/// # Arguments
///
/// * `file_path` - The path to the ELF file.
///
/// # Returns
///
/// Returns a `Result` containing the programming language used, if successfully determined.
/// Analysis example from: https://github.com/gimli-rs/gimli/blob/master/crates/examples/src/bin/simple.rs
pub fn dwarf_analysis(file_path: &str) -> Result<String>{
    let file = fs::File::open(&file_path)?;
    let mmap = unsafe { memmap2::Mmap::map(&file)? };
    let object = object::File::parse(&*mmap)?;
    let endian = if object.is_little_endian() {
        gimli::RunTimeEndian::Little
    } else {
        gimli::RunTimeEndian::Big
    };

    let lang = analyze_elf_file(&object, endian)?;
    Ok(lang.to_string())
}

// Parse the dwarf format in the .debug_info section. Language attributes table available here: https://dwarfstd.org/languages.html
fn analyze_elf_file<'b>(object: &'b object::File<'b>, endian: gimli::RunTimeEndian) -> Result<&'b str> {
    let load_section = |id: gimli::SectionId| -> Result<borrow::Cow<[u8]>> {
        match object.section_by_name(id.name()) {
            Some(ref section) => Ok(section
                .uncompressed_data()
                .unwrap_or(borrow::Cow::Borrowed(&[][..]))),
            None => Ok(borrow::Cow::Borrowed(&[][..])),
        }
    };
    let mut lang = "";
    let dwarf_cow = gimli::Dwarf::load(&load_section)?;
    let borrow_section: &dyn for<'a> Fn(
        &'a borrow::Cow<[u8]>,
    ) -> gimli::EndianSlice<'a, gimli::RunTimeEndian> =
        &|section| gimli::EndianSlice::new(&*section, endian);

    let dwarf = dwarf_cow.borrow(&borrow_section);
    let mut iter = dwarf.units();

    while let Some(header) = iter.next()? {
        let unit = dwarf.unit(header)?;
        let mut entries = unit.entries();

        while let Some((_, entry)) = entries.next_dfs()? {
            if let Some(language_attr) = entry.attr_value(gimli::DW_AT_language)? {
                let language = match language_attr {
                    gimli::AttributeValue::Language(language) => language,
                    _ => continue,
                };
                match language.static_string() {
                    Some(name) => {
                        if lang.contains(name){
                            return Ok(lang);
                        }
                        lang = name;
                    },
                    None => {},
                }
            }
        }
    }
    Ok(lang)
}
