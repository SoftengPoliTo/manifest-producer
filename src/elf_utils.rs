use std::{fs::File, io::Read};

use crate::error;
use capstone::prelude::*;
use error::{Error, Result};
use goblin::elf::{Elf, SectionHeader};
use object::elf::SHT_PROGBITS;


/* 
*
*   ELF Utilities: Functions for retrieving basic information from ELF files.
*
*/

// Structure used to collect API data identified in the code.
pub struct API {
    pub name: String,
    pub start_addr: u64,
    pub end_addr: u64,
    pub syscalls: Vec<String>,
}

impl API {
    pub fn new(name: String, start_addr: u64, end_addr: u64) -> Self {
        Self {
            name,
            start_addr,
            end_addr,
            syscalls: Vec::new(),
        }
    }

    pub fn add_syscall(&mut self, syscall: String) {
        self.syscalls.push(syscall);
    }
}

// Read the contents of an ELF file.
pub fn read_elf_file(file_path: &str) -> Result<Vec<u8>> {
    let mut file = File::open(&file_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;
    Ok(buffer)
}

// Check whether the specified ELF has been stripped of debug symbols.
pub fn is_stripped(elf: &Elf) -> bool {
    match elf.header.e_ident[goblin::elf::header::EI_CLASS] {
        goblin::elf::header::ELFCLASS64
        | goblin::elf::header::ELFCLASS32 => {
            !has_sections(&elf, goblin::elf::section_header::SHT_SYMTAB)
                || !has_sections(&elf, goblin::elf::section_header::SHT_STRTAB)
        }
        _ => true,
    }
}

fn has_sections(elf: &Elf, section_type: u32) -> bool {
    elf.section_headers.iter().any(|section| section.sh_type == section_type)
}

// Get the architecture type.
pub fn get_arch<'a>(elf: &'a Elf<'a>) -> Result<&'a str> {
    match elf.header.e_machine {
        goblin::elf::header::EM_X86_64 =>  Ok("x86-64"),
        goblin::elf::header::EM_386 =>  Ok("x86"),
        _ =>  Err(Error::InvalidElf { source: goblin::error::Error::Malformed("Unknown Architecture".to_string())}),
    }
}

// Return the file type.
pub fn get_file_type<'a>(elf: &'a Elf<'a>) -> Result<&'a str> {
    match elf.header.e_type {
        goblin::elf::header::ET_EXEC => Ok("Executable"),
        goblin::elf::header::ET_DYN => Ok("Dynamic Library"),
        goblin::elf::header::ET_CORE => Ok("File core"),
        _ => Err(Error::InvalidElf { source: goblin::error::Error::Malformed("Unknown File Type".to_string())}),
    }
}

// Check if the file is statically linked.
pub fn is_static(elf: &Elf) -> bool {
    if elf.dynamic.is_some() {
        false
    } else {
        true
    }
}

// Locate the .text section.
pub fn find_text_section<'a>(elf: &'a Elf<'a>) -> Option<&'a SectionHeader>{
    elf
        .section_headers
        .iter()
        .find(|sec| sec.sh_type == SHT_PROGBITS && {
            let name = elf.shdr_strtab.get_at(sec.sh_name);
            name == Some(".text")
        })
}

// Initialize Capstone.
pub fn cs_init() -> Result<Capstone> {
    let cs = Capstone::new()
        .x86()
        .mode(arch::x86::ArchMode::Mode64)
        .syntax(arch::x86::ArchSyntax::Att)
        .detail(true)
        .build();
    cs.map_err(|err| {
        Error::Capstone(format!("Failed to create Capstone instance: {}", err))
    })
}

// Retrieve the name given the address.
pub fn get_name_addr<'a>(elf: &'a Elf<'a>, address: u64) -> Option<&'a str> {
    let symtab = &elf.syms;
    let dyntab = &elf.dynsyms;
    if let Some(sym) = symtab.iter().find(|sym| sym.st_value == address) {
        if let Some(name) = elf.strtab.get_at(sym.st_name) {
            return Some(name);
        }
    } 
    else if let Some(dsym) = dyntab.iter().find(|dsym| dsym.st_value == address) {
        if let Some(name) = elf.dynstrtab.get_at(dsym.st_name) {
            return Some(name);
        }
    }
    else if let Some(text_section) = find_text_section(elf) {
        if let Some(name) = elf.strtab.get_at(text_section.sh_name) {
            return Some(name);
        }
    }
    None
}
