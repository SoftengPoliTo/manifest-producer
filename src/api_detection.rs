use goblin::elf::Elf;

use crate::{elf_utils, error};
use elf_utils::API;
use error::Result;

/// Do an API lookup in the symbol table.
///
/// This function searches for APIs in the symbol table of the ELF file based on a list of API names provided.
///
/// # Arguments
///
/// * `elf` - The ELF file structure.
/// * `api_list` - A vector containing the names of the APIs to search for.
///
/// # Returns
///
/// Returns a `Result` containing a vector of `API` structures representing the APIs found.
pub fn api_search<'a>(elf: &'a Elf<'a>, api_list: &'a [&'a str]) -> Result<Vec<API>> {
    let mut api_found = Vec::new();
    for symbol in &elf.syms {
        if symbol.st_type() == goblin::elf::sym::STT_FUNC && symbol.st_shndx != 0 {
            if let Some(function_name) = get_name_sym(elf, &symbol.to_owned()) {
                if api_list.contains(&function_name) {
                    api_found.push(API::new(
                        function_name.to_string(),
                        symbol.st_value,
                        symbol.st_value + symbol.st_size,
                    ));
                }
            }
        }
    }
    Ok(api_found)
}

// This function retrieves the name of a symbol from the ELF symbol table.
fn get_name_sym<'a>(elf: &'a Elf, symbol: &'a goblin::elf::Sym) -> Option<&'a str> {
    let name_offset = symbol.st_name;
    let name_str: &'a str = elf.strtab.get_at(name_offset)?;
    Some(name_str)
}
