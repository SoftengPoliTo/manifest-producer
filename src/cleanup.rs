use crate::{error, elf_utils};
use error::Result;

use elf_utils::API;
use cpp_demangle::{DemangleOptions, Symbol};
use rustc_demangle::demangle;

/* 
*
*   Cleanup: Functions for cleaning up and encapsulating system call flows.
*
*/

// Encapsulate the call flow within the appropriate structure.
pub fn syscall_flow(api: &mut API, sys: Vec<String>, lang: &str) -> Result<()>{
    for s in sys {
        if lang.contains("Rust") {
            if let Some(name) = clean_rust(&demangle_function_name(&s, true)?) {
                api.add_syscall(name);
            }
        }
        else {
            if let Some(name) = clean_cpp(&demangle_function_name(&s, false)?) {
                api.add_syscall(name);
            }
        }
    }
    Ok(())
}

// Attempt to clean up the mangled names
fn demangle_function_name(mangled_name: &str, rust: bool) -> Result<String> {
    if mangled_name.starts_with("_Z") {
        if rust {
            let demangled_name = demangle(mangled_name).to_string();
            return Ok(demangled_name);
        } 
        let options = DemangleOptions::default(); 
        let demangled_name = Symbol::new(mangled_name)?.demangle(&options)?; 
        Ok(demangled_name)
    } else {
        Ok(mangled_name.to_string())
    }
}

// Clean Rust call name
fn clean_rust(demangled_name: &str) -> Option<String> {
    let excluded_keywords = vec!["core::result", "shake_intern", "core::iter"];
    let contains_excluded = excluded_keywords.iter().any(|&keyword| demangled_name.contains(keyword));
    if contains_excluded {
        None
    } else {
        Some(demangled_name.to_string())
    }
}

// Clean C/C++ call name
fn clean_cpp(demangled_name: &str) -> Option<String> {
    let excluded_keywords = vec!["_Unwind_Resume", "shake_intern", "value_", "__cxa"];
    let contains_excluded = excluded_keywords.iter().any(|&keyword| demangled_name.contains(keyword));
    if contains_excluded {
        None
    } else {
        Some(demangled_name.to_string())
    }
}
