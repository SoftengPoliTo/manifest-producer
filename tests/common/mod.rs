use std::fs;
use std::path::Path;

use manifest_producer::api_detection::api_search;
use manifest_producer::cleanup::syscall_flow;
use manifest_producer::code_section_handler::code_section;
use manifest_producer::dwarf_analysis::dwarf_analysis;
use manifest_producer::elf_utils::{is_static, is_stripped, read_elf_file};
use manifest_producer::error::{Error, Result};
use manifest_producer::manifest_creation::{
    basic_info_manifest, feature_manifest, flow_call_manifest,
};

pub fn elf_analysis(file_path: &str, api_list: Vec<&str>, path: &str) -> Result<()> {
    let elf_data = read_elf_file(file_path)?;
    let elf = goblin::elf::Elf::parse(&elf_data)?;

    let stripped = is_stripped(&elf);
    if stripped {
        return Err(Error::DebugInfo);
    }

    let lang = match dwarf_analysis(file_path)?.strip_prefix("DW_LANG_") {
        Some(stripped_lang) => stripped_lang.to_owned(),
        None => "NOT_FOUND".to_string(), //return Err(Error::PrefixNotFound),
    };

    let link = is_static(&elf);

    let mut api_found = api_search(&elf, &api_list)?;
    if api_found.is_empty() {
        return Err(Error::APIListEmpty);
    }

    for api in &mut api_found {
        let sys = code_section(&elf, api, &elf_data, link, lang.contains("Rust"))?;
        syscall_flow(api, sys, &lang)?;
    }

    basic_info_manifest(&elf, file_path, &api_found, lang, path)?;
    flow_call_manifest(&api_found, path)?;
    feature_manifest(&api_found, path)?;

    Ok(())
}

pub fn compare_manifest(snapshot_path: &Path, path: &Path) {
    let content = fs::read_to_string(path).unwrap();
    let name = path.file_name().and_then(|v| v.to_str());
    insta::with_settings!({
        snapshot_path => snapshot_path,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, content);
    })
}
