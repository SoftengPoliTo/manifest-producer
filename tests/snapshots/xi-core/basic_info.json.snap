---
source: tests/common/mod.rs
expression: content
---
{
  "APIs found": [
    "get_flags"
  ],
  "architecture": "x86-64",
  "endianness": "Little",
  "entry_point": "0xc0bb0",
  "file_name": "xi-core",
  "file_type": "Dynamic Library",
  "header_size": 64,
  "link": "dynamically linked",
  "programming language": "Rust"
}
